File: CCC solutions/CCC '10/J5 - Knight Hop/J5 - Knight Hop.py
from collections import deque

change = [(1, 2), (2, 1), (2, -1), (1, -2), (-1, -2), (-2, -1), (-2, 1), (-1, 2)]
start = tuple(map(int, input().split(" ")))
end = tuple(map(int, input().split(" ")))
grid = {}


def graph(current):
    temp = []
    for i in range(1, 9):
        if current[0] + change[i - 1][0] > 8 or current[0] + change[i - 1][0] <= 0 or current[1] + change[i - 1][
            1] > 8 or current[1] + change[i - 1][1] <= 0:
            continue
        temp.append((current[0] + change[i - 1][0], current[1] + change[i - 1][1]))
    grid[current] = temp


def solve(start):
    qu = deque()
    qu.append(start)

    visited = {}
    prev = {}
    for x in range(1, 9):
        for y in range(1, 9):
            visited[(x, y)] = False
            prev[(x, y)] = None

    while len(qu) != 0:
        node = qu.popleft()
        neighbours = grid.get(node)

        for i in neighbours:
            if not visited[i]:
                graph(i)
                visited[i] = True
                qu.append(i)
                prev[i] = node
    return prev


def reconstructPath(start, end, prev):
    path = []
    currNode = prev[end]
    while currNode != start:
        path.append(currNode)
        currNode = prev[currNode]
    path.reverse()
    return path


def BFS(start, end):
    if start == end:
        return 0
    graph(start)
    prev = solve(start)
    return len(reconstructPath(start, end, prev)) + 1


print(BFS(start, end))

File: CCC solutions/CCC '10/J5 - Knight Hop/J5 - Knight Hop.cpp
#include <iostream>
#include <vector>
#include <algorithm>

typedef std::pair<int, int> p;
typedef std::vector<p> VEC;

VEC getNeighbors(p node)
{
	int dir[16] = { -2,1,  -1,2,  1,2,  2,1,  2,-1,  -1,-2,  -1,-2,  -2,-1 };
	VEC neighbors;
	for (int i = 0; i < 16; i += 2)
	{
		int x = node.second + dir[i + 1];
		int y = node.first + dir[i];
		if (x >= 1 && x <= 8 && y >= 1 && y <= 8)
			neighbors.push_back({ y,x });
	}return neighbors;
}

bool checkIn(std::vector<std::pair<p, bool>> arry, p child)
{
	for (auto a:arry)
		if (a.first == child)
			return true;
	return false;
}

VEC bfs(p start, p end)
{
	VEC qu = { start };
	std::vector<std::pair<p, p>> prev;
	std::vector<std::pair<p, bool>> visited;

	while (qu.size() != 0)
	{
		p node = qu[0];
		qu.erase(qu.begin());

		if (node == end)
		{
			p current = node;
			VEC path;
			while (current != start)
			{
				path.push_back(current);
				for (auto p: prev)
				{
					if (p.second == current)
					{
						current = p.first;
						break;
					}
				}
			}return path;
		}
		VEC neighbors = getNeighbors(node);

		for (auto child : neighbors)
		{
			if (!checkIn(visited, child))
			{
				visited.push_back({ child, true });
				qu.push_back(child);
				prev.push_back({ node,child });
			}
		}
	}
}

int main()
{
	p start;
	p end;

	std::cin >> start.first >> start.second;
	std::cin >> end.first >> end.second;

	VEC path = bfs(start, end);
	std::cout << path.size();
}